### DISCUSSION TITLE
Generating many mappings with python and json

### DISCUSSION THREAD

**Discussion by AlbertoV5** [helgoboss/helgobox#564] [Show and tell]
### Generating many mappings with python and json

Just adding this here for posterity, I haven't tried the new LUA mappings but for now, this is how I've been generating hundreds of mappings for TouchOSC.

Step 1. Creating a reference mapping in Realearn so source and target are configured as you want all the mappings to be set:

![0](https://user-images.githubusercontent.com/58243333/163531145-a6952f2f-23a2-487d-8b2a-71affca19313.JPG)

In this particular case, I am creating aliases for my FX Parameters and letting Realearn find the "Channel" alias FX and then the "LPF", "HPF"... etc aliases for FX Parameters. These are all the aliases I would use for whatever channelstrip I load:

```python

paramNames = [
    "LPF", "1QF", "2QF", "3QF", "4QF", 
    "HPF", "1FQ", "2FQ", "3FQ", "4FQ", 
    "INP", "1DB", "2DB", "3DB", "4DB",
    "CTR", "CRA", "CAT", "CRE", "CMO", 
    "GTR", "GRN", "GAT", "GRE", "GMO", 
    "OUT", "MIX", "PRE", "DRI", "BYP"]

```

I don't want to do it all by hand so:

Step 2. Copy the reference mapping and save it as .json:
![2](https://user-images.githubusercontent.com/58243333/163529195-a2aceab7-8ba8-441d-b025-a0bce35a752f.JPG)

Do it for as many variants you want to have, so if you are sending a float value, then a string formatted as {{target.text_value}} and another string formatted as {{target.fx_parameter.name}} you would need 3 different references because the there would be too many changes to the JSON file and it's easier to set it once on Realearn and then not touching the JSON much on the script.

You could do it without a reference but that's just to much JSON shenanigans so it's easier to start from a reference mapping.

Step 3. This is how the mapping would look, this example has the formatted string {{target.fx_parameter.name}}:

```json

{
  "kind": "Mapping",
  "value": {
    "id": "3Ta3MALJD322sJqkulzdI",
    "name": "name",
    "groupId": "BxdtiVP5sEuRAC2_hgMHD",
    "source": {
      "category": "osc",
      "isRegistered": false,
      "is14Bit": false,
      "oscAddressPattern": "/channel/track/1/fxparam/1/name",
      "oscArgIndex": 0,
      "oscArgType": "string"
    },
    "mode": {
      "maxStepSize": 0.05,
      "eelFeedbackTransformation": "{{target.fx_parameter.name}}",
      "feedbackColor": null,
      "feedbackType": 1
    },
    "target": {
      "trackExpression": "selected_track_index + (p[9] * 1000)",
      "fxAnchor": "name",
      "fxName": "Channel",
      "paramType": "name",
      "paramName": "LP",
      "useProject": true,
      "moveView": true,
      "seekPlay": true,
      "oscArgIndex": 0,
      "pollForFeedback": false
    }
  }
}

```

Then you save all the references as something like "value.json", "name.json" etc, then config and run the following python script:

### CODE HERE:
<details>

```python

  from pathlib import Path
  from uuid import uuid4
  import json
  
  
  class Mappings:
      """Make sure to change the mapping attributes and the internal method's json keys to fit your needs!
  
      Attributes
      ----------
      directory : Defaults to current working directory
      trackExpression : This is the dynamic track target to work with, leave {0} where the track index would be
      numberOfTracks : This is the number of times the same expression will repeat for different track indexes
      referenceJsonFiles : This is the original mapping reference you created in Realearn first
      targetFxName : Creating mappings for one plugin at a time
      paramNames : These are all the parameters you want to map, which repeat across tracks and references
      oscControl : {0} is track number, {1} is fxName, {2} is reference mapping / jsonFile
  
      Methods
      ----------
      modifyJson : change this method's json keys to fit your needs!
      """
  
      directory = Path.cwd()
      trackExpression = "selected_track_indexes[{0}]"
      numberOfTracks = 2
      
      referenceJsonFiles = ["value", "name", "str"]
      targetFxName = "Channel"
      paramNames = [
      "LPF", "1QF", "2QF", "3QF", "4QF", 
      "HPF", "1FQ", "2FQ", "3FQ", "4FQ", 
      "INP", "1DB", "2DB", "3DB", "4DB",
      "CTR", "CRA", "CAT", "CRE", "CMO", 
      "GTR", "GRN", "GAT", "GRE", "GMO", 
      "OUT", "MIX", "PRE", "DRI", "BYP"]
      oscControl = "/channel/track/{0}/fxparam/{1}/{2}"
  
  
      def modifyJson(self, jsonOutput, filePath, ref, trackNum, expression):
          control = ref
          for i in range(len(self.paramNames)):
  
              with open(filePath, "r") as file:
                  jsm = json.loads(file.read())["value"]
  
              jsm["id"] = str(uuid4())
              jsm["name"] = self.paramNames[i] + " {0}".format(control)
              jsm["source"]["oscAddressPattern"] = self.oscControl.format(str(trackNum), str(i+1), control)
  
              jsm["target"]["trackExpression"] = expression
              jsm["target"]["fxName"] = self.targetFxName
              jsm["target"]["paramName"] = self.paramNames[i]
  
              jsonOutput["value"].append(jsm)
  
          return jsonOutput
  
  
      def generate(self):
          """ construct a json output that contains ALL the mappings
          """
          jsonOutput = {"kind": "Mappings", "value":[]}   
          
          for i in range(self.numberOfTracks):
  
              for ref in self.referenceJsonFiles:
                  path = self.directory / (ref+".json")
                  jsonOutput = self.modifyJson(jsonOutput, path, ref, str(i+1), self.trackExpression.format(str(i)))
  
           with open (self.directory / "output.json", "w") as file:
               file.write(json.dumps(jsonOutput, indent=2))
  
  
  
  mappings = Mappings()
  
  mappings.directory = Path.cwd() / "scripts" / "realearn-python"
  
  mappings.generate()

```

</details>

What you want to change is the Attributes of your "mappings" instance as:

```python

mappings = Mappings()
mappings.paramNames = ["Threshold", "Ratio", "Attack", "Release"]
mappings.targetFxName = "VCS-1"
mappings.generate()

```

If you want to change how the JSON is addressed, just change the keys at modifyJson():

```python

jsm["target"]["fxName"] = self.targetFxName

```  

You can customize whatever target you want even after having created the reference mappings. As I said, I am using aliases for my FX and FX Param but you could use the real param names or whatever is your pleasure.

Step 4. The result will be in "output.json" and you can copy the contents of the file and then paste it into a new group in Realearn:

![4](https://user-images.githubusercontent.com/58243333/163530176-903fdd06-3a02-499b-a140-029ef5c38be0.JPG)

You can tweak and re-generate the output.json and replace all mappings in the group if you need to.

Then the TouchOSC side of things is another story but it works great:

![touchsoc](https://user-images.githubusercontent.com/58243333/163532455-8684d953-e5f0-449c-8f86-2c4b4cce7061.jpg)

Hope this made sense, haven't seen much of these shenanigans around.

This really works great with TouchOSC because the mappings are "infinite" so no matter what you want to control you can create hundreds of mappings and Realearn can handle it all really well. Make sure to enable/disble groups through Realearn parameters so you don't have thousands of mappings active at a time, but rather they get enabled as you page through your controller.

If y'all got any other ideas or ways of automating more of this stuff feel free to share them!

Thanks. Realearn is cool.

**Comment by helgoboss** [comment]
Cool stuff, thanks for this writeup.

BTW, even if you prefer Python over Lua, you might benefit from the new data format that came up with the introduction of ReaLearn Script (Lua). This new data format is just a plain data structure really, there's nothing Lua-specific in there! So you could easily use Python to generate a JSON that conforms to that new data format, ReaLearn will accept it. ReaLearn's integrated Lua import does more or less the same.

Here are some advantages of using the new format (directly written in JSON or generated by Python, Python, JavaScript or any language really):

- Much more self-explaining and consistent than ReaLearn's old format (the current native preset format). Reason: The old format is - well - very old. And it needs to be backward compatible until ReaLearn 1.0.
- The new format has an [auto-generated JSON Schema definition](https://github.com/helgoboss/realearn/blob/master/api/src/schema/generated/realearn.schema.json).
- I will eventually document it.
- At some point, the new format might becomes ReaLearn's preferred native preset format.

The advantage of specifically using Lua to generate data in this new format (optional, on top of that):

- It executes directly within ReaLearn, so you don't need anything else to generate the presets.
- At some point, ReaLearn will probably be able to save these Lua preset snippets, so they get "first class citizens".
- If I can convince Justin to expose ReaScript's Lua context as part of the REAPER SDK, one could get access to ReaScript functions when building presets (I can add partial support for this already now but I prefer not to because I don't want to recreate ReaScript).

**Reply by AlbertoV5** [reply]
Awesome! I'll slowly get into it. Having access to Reascript from Realearn would be legendary, even if it takes a while.

**Comment by vonglan** [comment]
I can confirm that generation with Lua is very easy, as is the language itself.

**Comment by tompad2** [comment]
Please could anyone make a Youtube tutorial or a written tutorial for a noob explaining how to make
presets this way! I am not good at python/json, a bit better at lua but no pro.
Everytime I hear people talking about making presets this way I wished I understood how to do this.
I have looked at exported code from ReaLearn but I dont get it.

**Reply by helgoboss** [reply]
Lua is the way to go IMO. It has the big advantage of being integrated into ReaLearn and is also supported as a native format for presets. With Python, you would always need an additional Python interpreter.

Can you make a small Lua export of your compartment (just a few mappings please) and ask more specific questions? What exactly is it that you don't get?

The export should be rather self-explaining, as it more or less looks like a "configuration file". It's Lua code, but there's no **for** loops or **if** conditions in there, so it shouldn't be that hard to understand. Or is exactly that the issue? That you don't know how to turn this into code with loops and conditions?

**Reply by tompad2** [reply]
_It's Lua code, but there's no for loops or if conditions in there, so it shouldn't be that hard to understand. Or is exactly that the issue? That you don't know how to turn this into code with loops and conditions?_

Exactly - I cant see how to do this.

This is connected to what I wrote in my other comment to you - sorry to split the conversation.

Its about both making presets for plugins and how to merge controller presets.
Does this make sense?

**Reply by tompad2** [reply]
Maybe we should only talk about presets for plugins here, and leave the other
thing in other chanel.

**Reply by helgoboss** [reply]
In that case, you just need to gain a bit Lua knowledge. How to introduce loops and conditions is not ReaLearn-specific.

Here's a super simple example, maybe you get an impression when you look at it: https://github.com/helgoboss/helgobox/blob/master/resources/main-presets/notes-to-clip-engine.lua

**Reply by tompad2** [reply]
Hmm, I have looked at your code, and I understand the for i = 0 to 127.
But sorry I dont understand how this code can get "connected" to a plugin?
It´s probably right in front of my eyes, but I cant see it.

Could you talk thru how I do if I have a plugin on a track and want to map all its parameters
to different buttons and encoders - this is maybe a subject for a Youtube tutorial? 
Or am I the only person not understanding :-)

**Reply by helgoboss** [reply]
By using target kind "FxParameterValue" instead?

The following script would create 8 mappings, one for each "Multi" (knob/fader/encoder), assuming that the controller preset exposes those as simple, numbered virtual control elements:

- Multi 0 will control parameter 0 of the first FX on the first track
- Multi 1 will control parameter 1 of the first FX on the first track
- Multi 2 will control parameter 2 of the first FX on the first track
- and so on ...

This is just for demonstration purposes. Completely up to you and your goal what you make from this. My suggestion is to import this script into ReaLearn instance and study/understand the result. Then change a target in the GUI to whatever you want. Then re-export the Lua for that single adjusted mapping by right-clicking the row => "Advanced => Copy as Lua (include default values)". Then adjust your script as desired.

And when you finally understood it, feel free to make a video tutorial :D I need help with those ;)

```lua
local mappings = {}

for i = 0, 8 do
    local mapping = {
        source = {
            kind = "Virtual",
            character = "Multi",
            id = i,
        },
        target = {
            kind = "FxParameterValue",
            parameter = {
                address = "ByIndex",
                fx = {
                    address = "ByIndex",
                    chain = {
                        address = "Track",
                        track = {
                            address = "ByIndex",
                            index = 0,
                        },
                    },
                    index = 0,
                },
                index = i,
            },
        },
    }
    table.insert(mappings, mapping)
end

return {
    kind = "MainCompartment",
    value = {
        mappings = mappings,
    },
}
```

**Reply by tompad2** [reply]
I am starting to see the light....

But, I am trying to make a script that create both Multi and Buttons.
Doing this didn't make it:
`source = {
            kind = "Virtual",
            character = "Multi",
            id = i,
             character = "Button",
            id = i,
        }`

Is there a way to do this or do I have to make 2 scripts, one for Multi and another for Button
and then merge them in texteditor?

**Reply by helgoboss** [reply]
None of that. I suppose you want to represent a push encoder, which can be said to be "Multi" and "Button" at the same time?

**Reply by tompad2** [reply]
Never mind!
I found out that putting a little AND at:
` character = "Multi" and "Button",`
made it!!


Thanks Benjamin! This made me understand all better.
Let see if I can make a video :-)
What program do you use to make ReaLearn-videos??

**Reply by tompad2** [reply]
Ooooh noooo - I didn't found out. I forgot to check
if Multis was listed too, they weren't....

**Reply by tompad2** [reply]
> None of that. I suppose you want to represent a push encoder, which can be said to be "Multi" and "Button" at the same time?

Yes, sort of. I have a XTouch Mini with both push encoders and buttons and a fader...

**Reply by helgoboss** [reply]
> Never mind! I found out that putting a little AND at: ` character = "Multi" and "Button",` made it!!

No, that can't work at all. Lua is not magic. With Lua import, you can only create things that you could also create via GUI. It's just a different way of getting there. The GUI also doesn't allow a virtual control element to be both a Multi AND Button.

> Thanks Benjamin! This made me understand all better. Let see if I can make a video :-) What program do you use to make ReaLearn-videos??

I use a screen recording tool for recording and REAPER for editing and export. But no pressure ;) Even a text tutorial in the wiki would be nice.

> Yes, sort of. I have a XTouch Mini with both push encoders and buttons and a fader...

Okay. A virtual control element can never be both "Button" and "Multi". If you have a push encoder, the "push" should be modelled as virtual control element with kind "Button". The encoder "turn" should be modelled as a separate virtual control element with kind "Multi".

Just to be clear:

- Those are the basics of building controller presets. This is not specific to Lua.
- The main preset then should match the virtual control elements defined in the controller preset.

**Reply by tompad2** [reply]
Ok, let see if I got it right - so I need to manually go in and learn source for the buttons?
Thats ok if so, its rather easy to do.

**Reply by helgoboss** [reply]
We really have to distinguish here between building controller presets and building main presets. If you don't know the difference, it's best to rewatch this video: https://www.youtube.com/watch?v=EqP3sKdVfj4

When building a controller preset, there's no real point of using Lua! Because it's not very repetitive and rather easy (see video). Lua can't help much there.

When building a main preset, Lua can be quite useful.

**Reply by tompad2** [reply]
I know the difference ( I have made the X-Touch mini controller preset).

It feels like I have not clearly explained what I am after.
Its a main preset I am building with the use of Lua-script you showed me.
Because the Lua-script can only list one sort of source from my controller (Multi or Button)
I need to edit the mappings that is no Multi by clicking Learn Source and touch the Button
to change its mapping to Button. That is the only way to get Multis and Buttons in 
mapping preset(main) right? Or am I wrong?

**Reply by helgoboss** [reply]
No, that's definitely wrong. To be honest, I don't really understand the problem that you are facing. Can you please post a screenshot or export of your desired main compartment mappings (just the essence, please)? Then I can show you how to achieve the same with Lua.

**Reply by tompad2** [reply]
:-) Ok, here comes a picture.

![Skärmklipp ReaLearn](https://github.com/user-attachments/assets/892fcad9-27bb-462a-b69e-3c3c7f9033aa)

The problem is, after importing from clipboard the lua-script I got from you earlier, all mappings
is Multi. Its ok for the first 18, but there is 32 buttons on controller (XTmini) that I want to map also.

I thought I had to select f.ex. Multi 19 mapping and click Learn Source to make it a Button 19 and do so
for every mapping I want to be Button, but you say that's definitely wrong ;-)

I cant see how I can do other way.

**Reply by helgoboss** [reply]
You mean like this!?

```lua
local mappings = {}

for i = 0, 8 do
    local mapping = {
        source = {
            kind = "Virtual",
            character = "Multi",
            id = i,
        },
        target = {
            kind = "FxParameterValue",
            parameter = {
                address = "ByIndex",
                fx = {
                    address = "ByIndex",
                    chain = {
                        address = "Track",
                        track = {
                            address = "ByIndex",
                            index = 0,
                        },
                    },
                    index = 0,
                },
                index = i,
            },
        },
    }
    table.insert(mappings, mapping)
end

for i = 0, 8 do
    local mapping = {
        source = {
            kind = "Virtual",
            character = "Button",
            id = i,
        },
        target = {
            kind = "FxParameterValue",
            parameter = {
                address = "ByIndex",
                fx = {
                    address = "ByIndex",
                    chain = {
                        address = "Track",
                        track = {
                            address = "ByIndex",
                            index = 0,
                        },
                    },
                    index = 0,
                },
                index = i,
            },
        },
    }
    table.insert(mappings, mapping)
end

return {
    kind = "MainCompartment",
    value = {
        mappings = mappings,
    },
}
```

**Reply by tompad2** [reply]
Hahahaa! Yes sort of :-)

Now I get 8 Rotarys and 8 RotaryPush and thats great!
But how do I get 8 ordinarie buttons? Do I need to make
another For-loop? 

Aaahaa - is this right? I make another For-loop from 9 to 17
to get the next ordinarie buttons!?

**Reply by tompad2** [reply]
Here we go!!

I think I got it - definitely?? ;-)

```
local mappings = {}

for i = 0, 16 do
    local mapping = {
        source = {
            kind = "Virtual",
            character = "Multi",
            id = i,
        },
        target = {
            kind = "FxParameterValue",
            parameter = {
                address = "ByIndex",
                fx = {
                    address = "ByIndex",
                    chain = {
                        address = "Track",
                        track = {
                            address = "ByIndex",
                            index = 0,
                        },
                    },
                    index = 0,
                },
                index = i,
            },
        },
    }
    table.insert(mappings, mapping)
end

for i = 0, 17 do
    local mapping = {
        source = {
            kind = "Virtual",
            character = "Button",
            id = i,
        },
        target = {
            kind = "FxParameterValue",
            parameter = {
                address = "ByIndex",
                fx = {
                    address = "ByIndex",
                    chain = {
                        address = "Track",
                        track = {
                            address = "ByIndex",
                            index = 0,
                        },
                    },
                    index = 0,
                },
                index = i,
            },
        },
    }
    table.insert(mappings, mapping)
end

for i = 18, 47 do
    local mapping = {
        source = {
            kind = "Virtual",
            character = "Button",
            id = i,
        },
        target = {
            kind = "FxParameterValue",
            parameter = {
                address = "ByIndex",
                fx = {
                    address = "ByIndex",
                    chain = {
                        address = "Track",
                        track = {
                            address = "ByIndex",
                            index = 0,
                        },
                    },
                    index = 0,
                },
                index = i,
            },
        },
    }
    table.insert(mappings, mapping)
end

return {
    kind = "MainCompartment",
    value = {
        mappings = mappings,
    },
}
```

**Reply by helgoboss** [reply]
Not sure if you got it. But if this yields the main compartment mappings that you were looking for, then you did at least something right ;)

They key points to understand ReaLearn's Lua support is:

- It's super simple, much simpler than ReaScript.
- All you need to do is to return a table of mappings (embedded into a parent table as shown in the `return` statement`). How you come up with those mappings is completely up to you.

In above example you are starting with an empty table of mappings. Then you gradually add mappings, at first 17, then 18 and so on.  Using `for` loops.
